connectors/utils.py

```diff
@@ -252,29 +255,65 @@ def _put(self, item):
         self._current_memsize += item[0]
         self._queue.append(item)
 
-    def mem_full(self):
-        if self.maxmemsize == 0:
-            return False
-        return self.qmemsize() >= self.maxmemsize
-
-    def qmemsize(self):
-        return self._current_memsize
+    def full(self, next_item_size=0):
+        full_by_numbers = super().full()
+        if full_by_numbers:
+            return True
+        return self._current_memsize + next_item_size >= self.maxmemsize
 
-    async def _wait_for_room(self, item):
-        item_size = get_size(item)
-        if self._current_memsize + item_size <= self.maxmemsize:
-            return item_size
+    async def _putter_timeout(self, putter):
+        """This coroutine will set the result of the putter to QueueFull when a certain timeout it reached."""
         start = time.time()
-        while self._current_memsize + item_size >= self.maxmemsize:
+        while not putter.done():
             if time.time() - start >= self.refresh_timeout:
-                raise asyncio.QueueFull()
+                putter.set_result(asyncio.QueueFull())
+                return
             logger.debug("Queue Full")
             await asyncio.sleep(self.refresh_interval)
-        return item_size
 
     async def put(self, item):
-        item_size = await self._wait_for_room(item)
-        return await super().put((item_size, item))
+        item_size = get_size(item)
+
+        # This block is taken from the original put() method but with two
+        # changes:
+        #
+        # 1/ full() takes the new item size to decide if we're going over the
+        #    max size, so we do a single call on `get_size` per item
+        #
+        # 2/ when the putter is done, we check if the result is QueueFull.
+        #    if it's the case, we re-raise it here
+        while self.full(item_size):
+            putter = self._get_loop().create_future()
+            putter_timeout = self._get_loop().create_task(self._putter_timeout(putter))
+            self._putters.append(putter)
+            try:
+                result = await putter
+                if isinstance(result, asyncio.QueueFull):
+                    raise result
+            except:  # NOQA
```

Comment:
I realize that this is the same code as in asyncio, but I would rather use `except Exception` so that Ctrl-C continues working. (It's unfortunate that asyncio does not care about Ctrl-C.)